File: examples/ios/Examples/Crud.m
// :replace-start: {
//   "terms": {
//     "CrudExampleObjc_": ""
//   }
// }
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>

// :code-block-start: models
// CrudExampleObjc_Person.h
@interface CrudExampleObjc_Person : RLMObject
@property NSNumber<RLMInt> *_id;
@property NSString *name;
@property (readonly) RLMLinkingObjects *dogs;
@end

// CrudExampleObjc_Dog.h
@interface CrudExampleObjc_Dog : RLMObject
@property NSString *name;
@property NSString *breed;
@property int age;
@property CrudExampleObjc_Person *owner;
@end

// CrudExampleObjc_Dog.m
@implementation CrudExampleObjc_Dog
@end

// CrudExampleObjc_Person.m
@implementation CrudExampleObjc_Person
+ (NSString *)primaryKey {
    return @"_id";
}
+ (NSDictionary *)linkingObjectsProperties {
    return @{
        @"dogs": [RLMPropertyDescriptor descriptorWithClass:CrudExampleObjc_Dog.class propertyName:@"owner"],
    };
}
@end
// :code-block-end:

@interface CrudObjc : XCTestCase

@end

@implementation CrudObjc

- (void)testBatchUpdateAndCascadingDelete {
    // :code-block-start: batch-update
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm transactionWithBlock:^() {
        // Create a person to take care of some dogs.
        CrudExampleObjc_Person *ali = [[CrudExampleObjc_Person alloc] initWithValue:@{@"_id": @1, @"name": @"Ali"}];
        [realm addObject:ali];

        // Find dogs younger than 2.
        RLMResults<CrudExampleObjc_Dog *> *puppies = [CrudExampleObjc_Dog objectsInRealm:realm where:@"age < 2"];

        // Batch update: give all puppies to Ali.
        [puppies setValue:ali forKey:@"owner"];
    }];
    // :code-block-end:

    CrudExampleObjc_Person *ali = [CrudExampleObjc_Person objectInRealm:realm forPrimaryKey:@1];
    // :code-block-start: cascading-delete
    [realm transactionWithBlock:^() {
        // Delete Ali's dogs.
        [realm deleteObjects:[ali dogs]];
        // Delete Ali.
        [realm deleteObject:ali];
    }];
    // :code-block-end:
}

- (void)testCreateAndDelete {
    // :code-block-start: create
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    // Instantiate the class.
    CrudExampleObjc_Dog *dog = [[CrudExampleObjc_Dog alloc] init];
    dog.name = @"Max";
    dog.age = 5;

    // Open a thread-safe transaction.
    [realm transactionWithBlock:^() {
        // Add the instance to the realm.
        [realm addObject:dog];
    }];
    // :code-block-end:

    // :code-block-start: delete
    [realm transactionWithBlock:^() {
        // Delete the instance from the realm.
        [realm deleteObject:dog];
    }];
    // :code-block-end:
}

- (void)testDeleteAll {
    // :code-block-start: delete-all
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm transactionWithBlock:^() {
        // Delete all objects from the realm.
        [realm deleteAllObjects];
    }];
    // :code-block-end:
}

- (void)testDeleteAllOfClass {
    // :code-block-start: delete-all-of-class
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm transactionWithBlock:^() {
        // Delete all instances of CrudExampleObjc_Dog from the realm.
        RLMResults<CrudExampleObjc_Dog *> *allCrudExampleObjc_Dogs = [CrudExampleObjc_Dog allObjectsInRealm:realm];
        [realm deleteObjects:allCrudExampleObjc_Dogs];
    }];
    // :code-block-end:
}

- (void)testDeleteCollection {
    // :code-block-start: delete-collection
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm transactionWithBlock:^() {
        // Find dogs younger than 2 years old.
        RLMResults<CrudExampleObjc_Dog *> *puppies = [CrudExampleObjc_Dog objectsInRealm:realm where:@"age < 2"];

        // Delete all objects in the collection from the realm.
        [realm deleteObjects:puppies];
    }];
    // :code-block-end:
}

- (void)testObjects {
    // :code-block-start: objects
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    RLMResults *dogs = [CrudExampleObjc_Dog allObjectsInRealm:realm];
    RLMResults *people = [CrudExampleObjc_Person allObjectsInRealm:realm];
    // :code-block-end:
    (void)dogs;
    (void)people;
}

- (void)testSort {
    // :code-block-start: sort
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    RLMResults *dogs = [CrudExampleObjc_Dog allObjectsInRealm:realm];

    // Sort dogs by name
    RLMResults *dogsSorted = [dogs sortedResultsUsingKeyPath:@"name" ascending:NO];

    // You can also sort on the members of linked objects. In this example,
    // we sort the dogs by dog's owner's name.
    RLMResults *ownersByName = [dogs sortedResultsUsingKeyPath:@"owner.name" ascending:YES];
    // :code-block-end:
    (void)dogsSorted;
    (void)ownersByName;
}

- (void)testFilter {
    // :code-block-start: filter
    RLMRealm *realm = [RLMRealm defaultRealm];

    // Access all dogs in the realm
    RLMResults *dogs = [CrudExampleObjc_Dog allObjectsInRealm:realm];

    // Filter by age
    RLMResults *puppies = [dogs objectsWhere:@"age < 2"];
    
    // Filter by owner
    RLMResults *availableDogs = [dogs objectsWhere:@"owner == nil"];
    
    // Filter by owner's name
    RLMResults *ownedByKeith = [dogs objectsWhere:@"owner.name == %@", @"Keith"];
    // :code-block-end:
    (void)puppies;
    (void)availableDogs;
    (void)ownedByKeith;
}

- (void)testTransaction {
    // :code-block-start: transaction
    // Open the default realm.
    RLMRealm *realm = [RLMRealm defaultRealm];

    // Open a thread-safe transaction.
    [realm transactionWithBlock:^() {
        // ... Make changes ...
        // Realm automatically cancels the transaction in case of exception.
        // Otherwise, Realm automatically commits the transaction at the
        // end of the code block.
    }];
    // :code-block-end:
}

- (void)testTransactionCounterexample {
    // :code-block-start: transaction-counterexample
    RLMRealm *realm = [RLMRealm defaultRealm];

    // BAD EXAMPLE -- avoid this!
    [realm beginWriteTransaction];
    // ... Make changes ...
    // If an exception is thrown here or the function otherwise returns early,
    // the transaction remains open!
    [realm commitWriteTransaction];
    // :code-block-end:
}

- (void)testUpdate {
    // :code-block-start: update
    RLMRealm *realm = [RLMRealm defaultRealm];
    // Open a thread-safe transaction.
    [realm transactionWithBlock:^() {
        // Get a dog to update.
        CrudExampleObjc_Dog *dog = [[CrudExampleObjc_Dog allObjectsInRealm: realm] firstObject];

        // Update some properties on the instance.
        // These changes are saved to the realm.
        dog.name = @"Wolfie";
        dog.age += 1;
    }];
    // :code-block-end:

}

- (void)testUpsert {
    // :code-block-start: upsert
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm transactionWithBlock:^() {
        CrudExampleObjc_Person *jones = [[CrudExampleObjc_Person alloc] initWithValue:@{@"ID": @1234, @"name": @"Jones"}];
        // Add a new person to the realm. Since nobody with ID 1234
        // has been added yet, this adds the instance to the realm.
        [realm addOrUpdateObject:jones];
        
        CrudExampleObjc_Person *bowie = [[CrudExampleObjc_Person alloc] initWithValue:@{@"ID": @1234, @"name": @"Bowie"}];
        // Judging by the ID, it's the same person, just with a different name.
        // This overwrites the original entry (i.e. Jones -> Bowie).
        [realm addOrUpdateObject:bowie];
    }];
    // :code-block-end:
}

@end

// :replace-end:

File: examples/ios/Examples/Crud.swift
// :replace-start: {
//   "terms": {
//     "CrudExample_": ""
//   }
// }
import XCTest
import RealmSwift

// :code-block-start: models
class CrudExample_Dog: Object {
    @objc dynamic var age = 0
    @objc dynamic var name = ""
    @objc dynamic var owner: CrudExample_Person?
}

class CrudExample_Person: Object {
    @objc dynamic var id = 0
    @objc dynamic var name = ""
    let dogs = LinkingObjects(fromType: CrudExample_Dog.self, property: "owner")

    override static func primaryKey() -> String? {
        return "id"
    }
}
// :code-block-end:

class Crud: XCTestCase {
    override func tearDown() {
        let realm = try! Realm()
        try! realm.write {
            realm.deleteAll()
        }
    }

    func testBatchUpdateAndCascadingDelete() {
        // :code-block-start: batch-update
        let realm = try! Realm()
        try! realm.write {
            // Create a person to take care of some dogs.
            let person = CrudExample_Person(value: ["id": 1, "name": "Ali"])
            realm.add(person)

            let dog = CrudExample_Dog(value: ["name": "Rex", "age": 1])
            realm.add(dog)

            // :hide-start:
            XCTAssert(person.dogs.count == 0)
            // :hide-end:
            // Find dogs younger than 2.
            let puppies = realm.objects(CrudExample_Dog.self).filter("age < 2")

            // Give all puppies to Ali.
            puppies.setValue(person, forKey: "owner")

            // :hide-start:
            XCTAssert(person.dogs.count == 1)
            // :hide-end:
        }
        // :code-block-end:

        // :code-block-start: cascading-delete
        let person = realm.object(ofType: CrudExample_Person.self, forPrimaryKey: 1)!
        // :hide-start:
        XCTAssert(person.dogs.count == 1)
        // :hide-end:
        try! realm.write {
            // Delete the related collection
            realm.delete(person.dogs)
            realm.delete(person)
        }
        // :code-block-end:
        XCTAssert(realm.objects(CrudExample_Dog.self).count == 0)
        XCTAssert(realm.objects(CrudExample_Person.self).count == 0)
    }

    func testCreateAndDelete() {
        // :code-block-start: create
        // Instantiate the class. For convenience, you can initialize
        // objects from dictionaries with appropriate keys and values.
        let dog = CrudExample_Dog(value: ["name": "Max", "age": 5])

        let realm = try! Realm()
        // Open a thread-safe transaction.
        try! realm.write {
            // Add the instance to the realm.
            realm.add(dog)
        }
        // :code-block-end:

        // :code-block-start: delete
        try! realm.write {
            // Delete the instance from the realm.
            realm.delete(dog)
        }
        // :code-block-end:
    }

    func testDeleteAll() {
        // :code-block-start: delete-all
        let realm = try! Realm()

        try! realm.write {
            // Delete all objects from the realm.
            realm.deleteAll()
        }
        // :code-block-end:
    }

    func testDeleteAllOfClass() {
        // :code-block-start: delete-all-of-class
        let realm = try! Realm()

        try! realm.write {
            // Delete all instances of Dog from the realm.
            let allDogs = realm.objects(CrudExample_Dog.self)
            realm.delete(allDogs)
        }
        // :code-block-end:
    }

    func testDeleteCollection() {
        // :code-block-start: delete-collection
        let realm = try! Realm()
        try! realm.write {
            // Find dogs younger than 2 years old.
            let puppies = realm.objects(CrudExample_Dog.self).filter("age < 2")

            // Delete the objects in the collection from the realm.
            realm.delete(puppies)
        }
        // :code-block-end:
    }

    func testObjects() {
        // :code-block-start: objects
        let realm = try! Realm()
        // Access all dogs in the realm
        let dogs = realm.objects(CrudExample_Dog.self)
        // :code-block-end:
    }

    func testSort() {
        // :code-block-start: sort
        let realm = try! Realm()
        // Access all dogs in the realm
        let dogs = realm.objects(CrudExample_Dog.self)

        let dogsSorted = dogs.sorted(byKeyPath: "name", ascending: false)

        // You can also sort on the members of linked objects. In this example,
        // we sort the dogs by dog's owner's name.
        let ownersByName = dogs.sorted(byKeyPath: "owner.name")
        // :code-block-end:
    }

    func testFilter() {
        // :code-block-start: filter
        let realm = try! Realm()
        // Access all dogs in the realm
        let dogs = realm.objects(CrudExample_Dog.self)

        // Filter by age
        let puppies = dogs.filter("age < 2")

        // Filter by owner
        let availableDogs = dogs.filter("owner == nil")

        // Filter by owner's name
        let ownedByKeith = dogs.filter("owner.name == 'Keith'")
        // :code-block-end:
        print(puppies.count)
        print(availableDogs.count)
        print(ownedByKeith.count)
    }

    func testTransaction() {
        // :code-block-start: transaction
        // Open the default realm.
        let realm = try! Realm()

        // Prepare to handle exceptions.
        do {
            // Open a thread-safe transaction.
            try realm.write {
                // Make any writes within this code block.
                // Realm automatically cancels the transaction
                // if this code throws an exception. Otherwise,
                // Realm automatically commits the transaction
                // after the end of this code block.
            }
        } catch let error as NSError {
            // Failed to write to realm.
            // ... Handle error ...
        }
        // :code-block-end:
    }

    func testTransactionCounterexample() {
        // :code-block-start: transaction-counterexample
        let realm = try! Realm()
        // BAD EXAMPLE -- avoid this!

        realm.beginWrite()
        // ... Make changes ...
        // If an exception is thrown before the commit,
        // this transaction stays open!
        try! realm.commitWrite()
        // :code-block-end:
    }

    func testUpdate() {
        let setupRealm = try! Realm()
        try! setupRealm.write {
            setupRealm.add(CrudExample_Dog())
        }
        // :code-block-start: update
        let realm = try! Realm()

        // Get a dog to update
        let dog = realm.objects(CrudExample_Dog.self).first!

        // Open a thread-safe transaction
        try! realm.write {
            // Update some properties on the instance.
            // These changes are saved to the realm
            dog.name = "Wolfie"
            dog.age += 1
        }
        // :code-block-end:
    }

    func testUpsert() {
        // :code-block-start: upsert
        let realm = try! Realm()
        try! realm.write {
            let person1 = CrudExample_Person(value: ["id": 1234, "name": "Jones"])
            // Add a new person to the realm. Since nobody with ID 1234
            // has been added yet, this adds the instance to the realm.
            realm.add(person1, update: .modified)

            let person2 = CrudExample_Person(value: ["id": 1234, "name": "Bowie"])
            // Judging by the ID, it's the same person, just with a
            // different name. When `update` is:
            // - .modified: update the fields that have changed.
            // - .all: replace all of the fields regardless of
            //   whether they've changed.
            // - .error: throw an exception if a key with the same
            //   primary key already exists.
            realm.add(person2, update: .modified)
        }
        // :code-block-end:
    }
}

// :replace-end:
